cli/daemon/run/runtime_config.go

```diff
@@ -0,0 +1,411 @@
+				if !found {
+					return "", errors.Newf("missing secret %s", secretName)
+				}
+			}
+		}
+	}
+
+	return encodeSecretsEnv(rtn), nil
+}
+
+// serviceConfigsForServices generates the service configs required for a service
+func (g *RuntimeEnvGenerator) serviceConfigsForServices(services []*meta.Service) (map[string]string, error) {
+	// Shortcut if we want it for all services
+	if len(services) == len(g.Meta.Svcs) {
+		return g.SvcConfigs, nil
+	}
+
+	// Otherwise build a map of just the configs this list of services need
+	rtn := make(map[string]string)
+	for _, service := range services {
+		if !service.HasConfig {
+			continue
+		}
+
+		cfg, found := g.SvcConfigs[service.Name]
+		if !found {
+			return nil, errors.Newf("missing computed concrete config for service %s", service.Name)
+		}
+
+		rtn[service.Name] = cfg
+	}
+
+	return rtn, nil
+}
+
+func (g *RuntimeEnvGenerator) init() {
+	g.initOnce.Do(func() {
+		g.deployID = fmt.Sprintf("run_%s", xid.New().String())
+		g.deployTime = time.Now()
+		g.authKey = g.AuthKey.GetOrElseF(genAuthKey)
+
+		g.pkgsBySvc = make(map[string][]*meta.Package)
+		for _, pkg := range g.Meta.Pkgs {
+			g.pkgsBySvc[pkg.ServiceName] = append(g.pkgsBySvc[pkg.ServiceName], pkg)
+		}
+
+		g.dbsBySvc = make(map[string][]*meta.SQLDatabase)
+		for _, svc := range g.Meta.Svcs {
+			for _, dbName := range svc.Databases {
+				for _, db := range g.Meta.SqlDatabases {
+					if db.Name == dbName {
+						g.dbsBySvc[svc.Name] = append(g.dbsBySvc[svc.Name], db)
+					}
+				}
+			}
+		}
+
+		g.topicsBySvc = make(map[string][]*meta.PubSubTopic)
+		g.subsBySvcByTopic = make(map[string]map[string][]*meta.PubSubTopic_Subscription)
+		for _, topic := range g.Meta.PubsubTopics {
+			for _, publisher := range topic.Publishers {
+				g.topicsBySvc[publisher.ServiceName] = append(g.topicsBySvc[publisher.ServiceName], topic)
+			}
+
+			for _, subscriber := range topic.Subscriptions {
+				g.topicsBySvc[subscriber.ServiceName] = append(g.topicsBySvc[subscriber.ServiceName], topic)
+
+				existing, found := g.subsBySvcByTopic[subscriber.ServiceName]
+				if !found {
+					existing = make(map[string][]*meta.PubSubTopic_Subscription)
+					g.subsBySvcByTopic[subscriber.ServiceName] = existing
+				}
+
+				g.subsBySvcByTopic[subscriber.ServiceName][topic.Name] = append(g.subsBySvcByTopic[subscriber.ServiceName][topic.Name], subscriber)
+			}
+		}
+
+		g.cachesBySvc = make(map[string][]*meta.CacheCluster)
+		for _, cacheCluster := range g.Meta.CacheClusters {
+			for _, keySpace := range cacheCluster.Keyspaces {
+				g.cachesBySvc[keySpace.Service] = append(g.cachesBySvc[keySpace.Service], cacheCluster)
+			}
+		}
+	})
+}
+
+// ListenAddresses is a list of listen address and port numbers for services to run on
+type ListenAddresses struct {
+	Gateway  netip.AddrPort            // The entrypoint to the application
+	Services map[string]netip.AddrPort // Map from service name to listen address
+}
```

Comment:
I kept this as it's own type, so we can use it as a type for managing service discovery map generation outside of the CLI too.

Hence the use of `netip.AddrPort` instead of `int16`, it allows for the option of the host being changed